course/include/isometry.h

```diff
@@ -96,5 +96,93 @@ inline std::ostream& operator<<(std::ostream& os, const Vector3& vector) {
               << ", z: " << std::to_string(static_cast<int>(vector.z())) << ")";
 }
 
+struct Rows {
+    Rows() : r1_(), r2_(), r3_() {};
+    Rows(const Vector3 &r1, const Vector3 &r2, const Vector3 &r3) : r1_(r1), r2_(r2), r3_(r3) {}; 
+    Vector3 r1_,r2_,r3_;
+};
+
+
+class Matrix3 {
+
+    public:
+        Matrix3() : m_(new Rows{}) {};
+
+        Matrix3(const Vector3 &r1, const Vector3 &r2, const Vector3 &r3) : m_(new Rows{r1, r2, r3}) {};
+        Matrix3(const std::initializer_list<double>& values);
+        ~Matrix3() {
+            // deallocate
+            delete m_;  
+        };
+
+        // copy constructor
+        Matrix3(const Matrix3& other) : m_(new Rows{*(other.m_)}) {};
+
+        // move constructor
+        Matrix3(Matrix3&& other) {
+            m_ = other.m_;
+            other.m_ = nullptr;
+        };
+
+        // Getter
+        const Vector3& row(uint32_t index) const;
+        const Vector3 col(uint32_t index) const;
+
+        // Getter for elements of the vector.
+        const Vector3& r1() const { return m_->r1_; };
+        const Vector3& r2() const { return m_->r2_; };
+        const Vector3& r3() const { return m_->r3_; };
+
+        // Setter for elements of the vector.
+        Vector3& r1() { return m_->r1_; };
+        Vector3& r2() { return m_->r2_; };
+        Vector3& r3() { return m_->r3_; };
+
+        // Operators overloading.
+        Vector3& operator[](const uint32_t index);
+        const Vector3& operator[](const uint32_t index) const;
+        Matrix3 operator+(const Matrix3& matrix) const;
+        Matrix3 operator-(const Matrix3& matrix) const;
+        Matrix3 operator*(const double& value) const;
+        Matrix3 operator*(const Matrix3& matrix) const;
+        Matrix3 operator/(const Matrix3& matrix) const;
+        Matrix3& operator+=(const Matrix3& matrix);
+        Matrix3& operator-=(const Matrix3& matrix);
+        Matrix3& operator*=(const double& value);
+        Matrix3& operator*=(const Matrix3& matrix);
+        Matrix3& operator/=(const Matrix3& matrix);
+        bool operator==(const Matrix3& matrix) const;
+        bool operator!=(const Matrix3& matrix) const;
+        Matrix3& operator=(const Matrix3& matrix);
+        Matrix3& operator=(Matrix3&& matrix);
+        double det() const;
+
+        // Constants
+        static const Matrix3 kIdentity;
+        static const Matrix3 kZero;
+        static const Matrix3 kOnes;
+    private:
+        Rows *m_;
```

Comment:
No need to change it now, but notice that since you already implemented the move operation in `Vector3`, you don't need to use pointers here to reimplement move here (same goes for Isometry). You could just define the three rows, and whenever you need to move the matrix you'd do for instance:
```
row[0] = std::move(other.row[0]);
row[1] = std::move(other.row[1]);
row[2] = std::move(other.row[2]);
```